### ISSUE TITLE
Custom layer that calls non-python code

### ISSUE THREAD

**Issue by adler-j** [fchollet/keras#5517]
I am working on a project where I need to implement a custom layer that applies some transformation to my data, with the caveat is that that transformation is implemented in non-python code that cannot be used with the auto-differentiation used by Keras. With that said, I can calculate the gradients, and simply need a way to let Keras know it should use my supplied gradients instead of trying to use auto-differentiation.

In tensorflow, it seems that this can be solved by [registering a gradient implementation](https://www.tensorflow.org/extend/adding_an_op#implement_the_gradient_in_python). Is there something similar that I could do in Keras?

**Comment by unrealwill** [comment]
Hello,

Keras is fundamentally a wrapper of theano/tensorflow operations.
So you will need to implement your operation in theano/tensorflow, then you create a Layer which wraps your operation.

**Comment by adler-j** [comment]
Thank you for the input! Is there any example of how to wrap a tensorflow layer in Keras?

**Comment by unrealwill** [comment]
Look at :  https://keras.io/layers/writing-your-own-keras-layers/
And eventually https://github.com/fchollet/keras/blob/master/keras/backend/tensorflow_backend.py

**Comment by adler-j** [comment]
So, if there is no official example of this, would it be something that the Keras community would be interested in me adding? I guess I'm not the first nor last one who wants to do something like this.

**Comment by nayansinghal** [comment]
I am also facing the same issue. Did you solve it and if yes, would it be possible to provide a link to your code?

**Comment by adler-j** [comment]
Basically I ended up swapping to tensorflow where this is supported natively, see e.g. these projects that use it:

https://github.com/adler-j/learned_primal_dual
https://github.com/adler-j/wasserstein_inverse_problems
https://github.com/adler-j/learned_gradient_tomography

**Comment by nayansinghal** [comment]
Do you mean that there is no good way to write the layer which can wrap around custom tensorflow ops?

I want a behavior in which during the forward pass, it will do quantization and for the backward pass, it will behave as an identity and pass the gradients as it is without considering the gradient of that layer.
Is there any method in Keras which can solve this issue?

**Comment by fchollet** [comment]
Well, if you can do it in TF, obviously you can do it in Keras (through TF).

**Comment by nayansinghal** [comment]
I am trying but I am getting error while training. 
https://github.com/fchollet/keras/issues/8526

Can you please help me with it?

**Comment by fchollet** [comment]
Does your code work in TF, without Keras? The error you're getting is
TF-side.

**Comment by nayansinghal** [comment]
issue resolved. It was happening because we got numpy object in quantization() and I was applying tensor functions on it.
